openedx_proversity_reports/reports/last_page_accessed.py

```diff
@@ -0,0 +1,183 @@
+"""
+Last page accessed reports.
+"""
+from completion.models import BlockCompletion
+from django.contrib.auth.models import User
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey
+
+from openedx_proversity_reports.edxapp_wrapper.get_course_blocks import get_course_blocks
+from openedx_proversity_reports.edxapp_wrapper.get_modulestore import get_modulestore
+from openedx_proversity_reports.edxapp_wrapper.get_student_library import user_has_role, get_course_staff_role
+
+
+def get_last_page_accessed_data(course_list):
+    """
+    Returns a dict with information about the last page accessed
+    by the student according to completion model.
+
+    {
+        'course-id': [{
+            'username': User name,
+            'last_time_accessed': Date string,
+            'last_page_viewed': String of the problem's parent block tree,
+            'block_id': Block id of the lastest block accessed by the student,
+            'vertical_block_id': Parent vertical block id,
+        }]
+    }
+    """
+    last_page_data = {}
+
+    for course_id in course_list:
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            continue
+
+        # Getting all students enrolled in the course except staff users
+        enrolled_students = User.objects.filter(
+            courseenrollment__course_id=course_key,
+            courseenrollment__is_active=1,
+            is_staff=0,
+        )
+
+        if not enrolled_students:
+            continue
+
+        user_data = []
+
+        usage_key = get_modulestore().make_course_usage_key(course_key)
+        blocks = get_course_blocks(enrolled_students[0], usage_key)
+
+        for user in enrolled_students:
+            if user_has_role(user, get_course_staff_role(course_key)):
+                continue
+            last_completed_child_position = BlockCompletion.get_latest_block_completed(user, course_key)
+            parent_tree_name = ''
+
+            if last_completed_child_position:
+                vertical_blocks = blocks.topological_traversal(
+                    filter_func=lambda block_key: block_key.block_type == 'vertical',
+                    yield_descendants_of_unyielded=True,
+                )
+
+                for vertical in vertical_blocks:
+                    for problem in blocks.get_children(vertical):
+                        if problem.block_id == last_completed_child_position.block_key.block_id:
+                            parent_tree = get_parent_tree(blocks, problem)
```

Comment:
get_parent_tree can be called inside get_parent_display_names

Reply:
Addressed.